src/models/blockchain/block/pow.rs

```diff
@@ -425,4 +608,248 @@ pub(crate) mod tests {
+                .await
+                .expect("preprocessing should have aborted by now");
+
+            // assert not too much time passed
+            let elapsed = tick.elapsed();
+            assert!(
+                elapsed < Duration::from_secs(1),
+                "elapsed time between channel-cancellation and task finish: {:?} > 1 s",
+                elapsed
+            );
+
+            // assert that preprocessing was indeed aborted
+            assert_eq!(guesser_buffer, GuesserBuffer::default());
+        }
+
+        #[apply(shared_tokio_runtime)]
+        async fn can_cancel_merkle_tree_construction_within_two_seconds() {
+            const MERKLE_TREE_HEIGHT: usize = 29;
+            let num_leafs = 1 << MERKLE_TREE_HEIGHT;
+            let (cancel_tx, mut cancel_rx) = oneshot::channel::<()>();
+            let (start_tx, mut start_rx) = oneshot::channel::<()>();
+
+            // spawn preprocess task
+            let join_handle = tokio::task::spawn_blocking(move || {
+                let leafs = vec![Digest::default(); num_leafs];
+                let internal_nodes = leafs.clone();
+                start_tx.send(()).unwrap();
+                MTree::build_inplace(leafs, internal_nodes, Some(&cancel_tx))
+            });
+
+            // wait until we are sure the task started
+            while start_rx.try_recv().unwrap().is_none() {
+                tokio::time::sleep(Duration::from_millis(10)).await;
+            }
+
+            // cancel channel
+            cancel_rx.close();
+            let tick = Instant::now();
+
+            // wait until task is finished
+            let mtree = join_handle
+                .await
+                .expect("preprocessing should have aborted by now");
+
+            // assert not too much time passed
+            let elapsed = tick.elapsed();
+            assert!(
+                elapsed < Duration::from_secs(2),
+                "elapsed time between channel-cancellation and task finish: {:?} > 2 s",
+                elapsed
+            );
+
+            // assert that merkle tree construction did not finish
+            assert_eq!(mtree.num_leafs(), 0);
+        }
+    }
+
+    mod merkle_tree_tests {
+        use super::*;
+        use rand::rngs::StdRng;
+        use rand::SeedableRng;
+        use tasm_lib::twenty_first::prelude::MerkleTree;
+
+        #[test]
+        fn roots_agree() {
+            let mut rng = rng();
+            let height = 10;
+            let num_leafs = 1 << height;
+            let leafs = (0..num_leafs).map(|_| rng.random::<Digest>()).collect_vec();
+            let merkle_tree =
+                MerkleTree::sequential_new(&leafs).expect("must not forget to unwrap");
+            let internal_nodes = vec![Digest::default(); num_leafs];
+            let mtree = MTree::build_inplace(leafs, internal_nodes, None);
+
+            assert_eq!(mtree.root(), merkle_tree.root());
+        }
+
+        #[test]
+        fn paths_agree() {
```

Comment:
Suggestion: Rename to `authentication_paths` agree, to highlight that we're not just testing indices.